README.md
# bash-inline

## Introduction

`bash-inline` is a CLI tool to format Bash scripts for inlining.

## Usage

In order to format a Bash script to be readable you can run `bash-inline` without any further paramters.
By default `bash-inline` reads your Bash script from standard input to standard output.

```
# Given a Bash script like:
# 
# | #!/usr/bin/env bash
# | echo "Is true just 1?"
# | 
# | # We'll see
# | if [[ true -eq 1 ]]; then
# |   echo "True is really just 1"
# | fi
# 
# Running bash-inline on it will remove comments, format (e.g. add missing semicolons) and check your script:
# 
# | echo "Is true just 1?";
# | if [[ true -eq 1 ]]; then
# |   echo "True is really just 1";
# | fi;
#
# This can be also used for making a one-liner more readable.
cat example.sh | bash-inline
```

The same script can also be one-lined when passing the `-1` flag.

```bash
# Running bash-inline with the -1 flag will one-line it:
#
# | echo "Is true just 1?"; if [[ true -eq 1 ]]; then echo "True is really just 1"; fi;
#
cat example.sh | bash-inline -1
```

You can also read and output the Bash script from a file, and pass some arguments for formatting it given some line length and ribbon ratio constraints. 
For more information on how to use `bash-inline` run `bash-inline --help`

## Building

This project is built with [GHC](http://www.haskell.org/ghc/) and [cabal-install](http://hackage.haskell.org/package/cabal-install), you can get the latest version by installing the [Haskell platform](http://hackage.haskell.org/platform/).

Update dependencies and configure the project with tests enabled:
```bash
cabal update 
cabal configure --enable-tests
```

Build the project and run tests:
```bash
cabal build
cabal test
```

If you want to install the latest version of `bash-inline`:
```bash
# Be sure that the $CABALDIR/bin directory is in your path
cabal install
```

## Special Thanks

To everyone who contributed to [langugae-bash](http://hackage.haskell.org/package/language-bash) and [pretty](http://hackage.haskell.org/package/pretty) which provides the core functionality to this tool.



ChangeLog.md
# Revision history for bash-inline

## 0.3.0.0  -- 2019-10-22

* Third release. It's getting better.
* Use the rendering features of pretty to make rendering of one-liners more robust. 
* Add additional paramters for controlling the maximum line length, the ribbon ratio and outputting the script.

## 0.2.0.1  -- 2019-10-16

* Bugfix newline escaping in error output.

## 0.2.0.0  -- 2019-10-16

* Second release. API already broken.
* Pass in a Bash script from a file using an option.
* Add flag for pretty printing Bash scripts on one line.

## 0.1.0.0  -- 2019-10-15

* First version. Released on an unsuspecting world.
* Parse and pretty print Bash scripts from a file or stdin.


test/cond.sh
#!/usr/bin/env bash
if [[ true -eq 1 ]]; then
	echo "true"
else
	echo "false"
fi

if [[ false -eq 1 ]]; then
	echo "false"
else
	echo "true"
fi

if [[ false -eq 1 ]]; then
	echo "false"
elif [[ false -eq 0 ]]; then
	echo "true"
else
	echo "false"
fi

if [[ false -eq 1 ]]; then
	echo "false"
elif [[ true -eq 0 ]]; then
	echo "false"
else
	echo "true"
fi


test/func.sh
#!/usr/bin/env bash
get_name() {
  echo "John"
}

echo "You are $(get_name)"


test/vars.sh
#!/usr/bin/env bash
NAME="John"
echo $NAME
echo "$NAME"
echo "${NAME}!"


src/Lib.hs
module Lib (
  Mode(..), LineLength, RibbonsPerLine, Script,
  eitherParse,
  format
) where

import           Data.Bifunctor
import           Data.Maybe
import qualified Language.Bash.Syntax as Bash
import qualified Language.Bash.Parse  as Bash
import qualified Language.Bash.Pretty as Bash
import qualified Text.Parsec.Error    as Parsec              
import qualified Text.Parsec.Pos      as Parsec
import qualified Text.PrettyPrint     as PrettyPrint

-- | Formatting mode, either normal or one-lined output
data Mode = Normal | OneLine

-- | Maximum line length of the formatted output
type LineLength = Maybe Int

-- | Maximum ratio of line length to characters on a line excluding indentation
type RibbonsPerLine = Maybe Float

-- | A compount list of statements
type Script = Bash.List

-- | Parse a Bash script, return an error message when parsing fails.
eitherParse :: String -> Either String Script 
eitherParse content = 
  let sourceName = "<ignored>"
  in  first mapErrorToString $ Bash.parse sourceName content
  where
    mapErrorToString :: Parsec.ParseError -> String
    mapErrorToString e = 
      "error " ++ showSourcePos (Parsec.errorPos e) ++ ":" ++ showMessages (Parsec.errorMessages e)

    showSourcePos :: Parsec.SourcePos -> String
    showSourcePos pos =
      let line   = show $ Parsec.sourceLine   pos
          column = show $ Parsec.sourceColumn pos
      in  "(line " ++ line ++ ", column " ++ column ++ ")"
    
    showMessages :: [Parsec.Message] -> String
    showMessages =
      Parsec.showErrorMessages "or" "unknown parse error" "expecting" "unexpected" "end of input"

-- | Format and render a Bash script to a string
format :: Mode -> LineLength -> RibbonsPerLine -> Script -> String
format m l r script = 
  let postamble = ""
  in  PrettyPrint.fullRender (mode m) (lineLength l) (ribbonsPerLine r) printer postamble (Bash.pretty script)
  where
    mode :: Mode -> PrettyPrint.Mode
    mode Normal  = PrettyPrint.PageMode
    mode OneLine = PrettyPrint.OneLineMode

    lineLength :: LineLength -> Int
    lineLength = fromMaybe 80

    ribbonsPerLine :: RibbonsPerLine -> Float
    ribbonsPerLine = fromMaybe 1.5

    printer :: PrettyPrint.TextDetails -> String -> String
    printer (PrettyPrint.Chr  c) s  = c :  s
    printer (PrettyPrint.Str  s) s' = s ++ s'
    printer (PrettyPrint.PStr s) s' = s ++ s'


app/Main.hs
module Main where

import Lib
import Paths_bash_inline (version)

import Data.Semigroup ((<>))
import Data.Version (showVersion)
import qualified Options.Applicative  as Opts
import System.Exit
import System.IO           

data Args = Args {
    argFile           :: Maybe FilePath
  , argOutput         :: Maybe FilePath
  , argOneLine        :: Bool
  , argLineLength     :: LineLength
  , argRibbonsPerLine :: RibbonsPerLine 
  } 

runWithArgs :: Args -> IO ()
runWithArgs (Args maybeInPath maybeOutPath oneLine lineLength ribbonsPerLine) = do
  src <- maybe getContents readFile maybeInPath 
  case eitherParse src of
    Left  err  -> putStrLn err >> exitFailure
    Right src' -> output maybeOutPath $ format (modeOf oneLine) lineLength ribbonsPerLine src'
  where
    output :: Maybe FilePath -> String -> IO ()
    output Nothing     src = putStrLn src
    output (Just path) src = writeFile path $ "#!/usr/bin/env bash\n" ++ src

    modeOf :: Bool -> Mode
    modeOf True  = OneLine
    modeOf False = Normal

main :: IO ()
main = Opts.execParser optsParser >>= runWithArgs
  where
    optsParser :: Opts.ParserInfo Args
    optsParser = Opts.info (Opts.helper <*> versionOpt <*> programOpts) $
         Opts.fullDesc 
      <> Opts.header "bash-inline - format a Bash script for inlining" 

    versionOpt :: Opts.Parser (a -> a)
    versionOpt = Opts.infoOption ("bash-inline " ++ showVersion version) $ 
         Opts.long "version" 
      <> Opts.short 'v' 
      <> Opts.help "Output version information and exit"

    oneLineOpt :: Opts.Parser Bool
    oneLineOpt = Opts.switch $
          Opts.long "one-line" 
       <> Opts.short '1'
       <> Opts.help "Output script in one line"

    lineLengthOpt :: Opts.Parser LineLength
    lineLengthOpt = Opts.optional . Opts.option Opts.auto $
          Opts.long "lineLength"
       <> Opts.short 'l'
       <> Opts.help "Maximum line length of the formatted output [80]"

    ribbonsPerLineOpt :: Opts.Parser RibbonsPerLine
    ribbonsPerLineOpt = Opts.optional . Opts.option Opts.auto $
          Opts.long "ribbonsPerLine"
       <> Opts.short 'r'
       <> Opts.help "Maximum ratio of line length to characters on a line excluding indentation [1.5]"

    fileOpt :: Opts.Parser (Maybe FilePath)
    fileOpt = Opts.optional . Opts.strOption $
          Opts.long "file" 
       <> Opts.short 'f' 
       <> Opts.help "Bash script file to read instead of standard input"

    outputOpt :: Opts.Parser (Maybe FilePath)
    outputOpt = Opts.optional . Opts.strOption $
          Opts.long "output"
       <> Opts.short 'o'
       <> Opts.help "Bash script file to output the formatted script"

    programOpts :: Opts.Parser Args
    programOpts = Args <$> fileOpt <*> outputOpt <*> oneLineOpt <*> lineLengthOpt <*> ribbonsPerLineOpt 
